src/Likvido.ApiAuth.Common/UserInfoHeaderSerializer.cs

```diff
@@ -0,0 +1,135 @@
+                }
+            }
+
+            if (!hasId)
+            {
+                claims.Add(new Claim(ClaimTypes.NameIdentifier, jsonUserId));
+            }
+
+            return new ClaimsIdentity(claims, authType);
+        }
+
+        public static string  Serialize(string userId, IEnumerable<Claim> claims, TokenClaims tokenClaims = null)
+        {
+            if (string.IsNullOrWhiteSpace(userId))
+            {
+                throw new ArgumentException("Value muct be provided", userId);
+            }
+
+            var idClaim = claims.SingleOrDefault(c => c.Type == ClaimTypes.NameIdentifier);
+            if (idClaim != null && idClaim.Value != userId)
+            {
+                throw new ArgumentException($"UserId and \"{ClaimTypes.NameIdentifier}\" value must match", nameof(userId));
+            }
+
+            tokenClaims = tokenClaims ?? TokenClaims.Default;
+
+            var userInfo = new ApiUserInfo
+            {
+                UserId = userId
+            };
+
+            var claimGroups = claims.GroupBy(c => c.Type).ToList();
+            foreach (var claimGroup in claimGroups)
+            {
+                var claimName = tokenClaims.FromClaimToTokenClaim(claimGroup.Key);
+                if (claimGroup.Count() > 1)
+                {
+                    userInfo.Claims.Add(claimName, claimGroup.Select(c => c.Value).ToArray());
+                }
+                else
+                {
+                    userInfo.Claims.Add(claimName, claimGroup.First().Value);
+                }
+            }
+
+            return Convert.ToBase64String(Encoding.UTF8.GetBytes(JsonSerializer.Serialize(userInfo, SerializerOptions)));
+        }
+
+        public static string  Serialize(IEnumerable<Claim> claims, TokenClaims tokenClaims = null)
+        {
+            var idClaim = claims.SingleOrDefault(c => c.Type == ClaimTypes.NameIdentifier);
+            if (string.IsNullOrWhiteSpace(idClaim?.Value))
+            {
+                throw new ArgumentException($"\"{ClaimTypes.NameIdentifier}\" claim must exist", nameof(claims));
+            }
+
+            tokenClaims = tokenClaims ?? TokenClaims.Default;
+            var userInfo = new ApiUserInfo
+            {
+                UserId = idClaim.Value
+            };
+
+            var claimGroups = claims.GroupBy(c => c.Type).ToList();
+            foreach (var claimGroup in claimGroups)
+            {
+                var claimName = tokenClaims.FromClaimToTokenClaim(claimGroup.Key);
+                if (claimGroup.Count() > 1)
+                {
+                    userInfo.Claims.Add(claimName, claimGroup.Select(c => c.Value).ToArray());
+                }
+                else
+                {
+                    userInfo.Claims.Add(claimName, claimGroup.First().Value);
+                }
+            }
+
+            return Convert.ToBase64String(Encoding.UTF8.GetBytes(JsonSerializer.Serialize(userInfo, SerializerOptions)));
```

Comment:
Maybe extract this into a private re-usable method?

Reply:
I forgot to reuse the whole `Serialize`. Now fixed